gsuite_reports_rules/gsuite_drive_visibility_change.py

```diff
@@ -1,42 +1,123 @@
         ):
-            return True
+            global TARGET_USER_EMAIL
+            TARGET_USER_EMAIL = param_lookup(details.get("parameters", {}), "target_user")
+            DOC_TITLE = param_lookup(details.get("parameters", {}), "doc_title")
+            ACCESS_SCOPE = param_lookup(details.get("parameters", {}), "new_value")
+
+            change_user_access = True
+            break
+
+    if change_user_access:
+        return True
 
     return False
 
 
 def dedup(event):
-    doc_title = "<UNKNOWN_DOC_TITLE>"
-    for detail in event.get("events", [{}]):
-        if detail.get("type") == "acl_change":
-            if param_lookup(detail.get("parameters", {}), "doc_title"):
-                doc_title = param_lookup(detail.get("parameters", {}), "doc_title")
-                break
-    return doc_title
+    return DOC_TITLE
 
 
 def title(event):
-    target_user_email = "<EMAIL_UNKNOWN>"
-    doc_title = "<UNKNOWN_DOC_TITLE>"
-    for detail in event.get("events", [{}]):
-        if detail.get("type") == "acl_change":
-            if param_lookup(detail.get("parameters", {}), "doc_title"):
-                doc_title = param_lookup(detail.get("parameters", {}), "doc_title")
-            if param_lookup(detail.get("parameters", {}), "target_user"):
-                target_user_email = param_lookup(detail.get("parameters", {}), "target_user")
-            break
+
+    if TARGET_USER_EMAIL != "<UNKNOWN_USER>":
+        sharing_scope = TARGET_USER_EMAIL
+        # TODO: make sure this options only appears if a target_user is specified
+        if NEW_VISIBILITY == "shared_externally":
+            sharing_scope = sharing_scope + " (outside the document's current domain)"
+    # TODO: make sure there is no target_domain if there is a target_user
+    elif TARGET_DOMAIN == "all":
+        sharing_scope = "the entire internet"
+        # TODO: make sure these options only appear with the "all" target_domain
+        if NEW_VISIBILITY == "people_with_link":
+            sharing_scope = sharing_scope + " (anyone with the link)"
+        elif NEW_VISIBILITY == "public_on_the_web":
+            sharing_scope = sharing_scope + " (link not required)"
+    else:
+        sharing_scope = f"the {TARGET_DOMAIN} domain"
+        # TODO: make sure these options only appear with a specific target_domain (not "all")
+        if NEW_VISIBILITY == "people_within_domain_with_link":
+            sharing_scope = sharing_scope + f" (anyone in {TARGET_DOMAIN} with the link)"
+        elif NEW_VISIBILITY == "public_in_the_domain":
+            sharing_scope = sharing_scope + f" (anyone in {TARGET_DOMAIN})"
+
+    global ACCESS_SCOPE
+    # TODO: confirm multiValue always has one element
+    if ACCESS_SCOPE == ["can_view"]:
+        ACCESS_SCOPE = "view"
+    elif ACCESS_SCOPE == ["can_comment"]:
+        ACCESS_SCOPE = "comment"
+    elif ACCESS_SCOPE == ["can_edit"]:
+        ACCESS_SCOPE = "edit"
```

Comment:
I feel like we can either do something clever or do something clear here, but I don't find this clever or clear.

Something clever (assuming we don't know if `ACCESS_SCOPE` is always a list, if it's always a list we can drop the `isinstance` part but the default value as you have it is a string):

```python
alert_access_scope = ACCESS_SCOPE[0].strip("can_") if ACCESS_SCOPE and isinstance(ACCESS_SCOPE, list) else ACCESS_SCOPE
```

Something clear:
```python
alert_access_scope = ACCESS_SCOPE
if ACCESS_SCOPE and isinstance(ACCESS_SCOPE, list):
  alert_access_scope = ACCESS_SCOPE[0].trim("can_")
```

In either case, we should ditch `global` here since we have no need to modify this global variable.

Reply:
Yes, it is always a list. Actually, don't even need to test `if ACCESS_SCOPE`, because it will always be there if the rule is triggered. Switching to `alert_access_scope = ALERT_DETAILS["ACCESS_SCOPE"][0].trim("can_")`